### DISCUSSION TITLE
Specify name of #[pyclass] dynamically in macro

### DISCUSSION THREAD

**Discussion by rcrisanti** [PyO3/pyo3#4612] [Questions]
I'm attempting to use a macro (with the help of [paste](https://docs.rs/paste/latest/paste/)) to make a class inheritance hierarchy less verbose. What I would ideally like to do is something like this:
```rust
use paste::paste;
use pyo3::prelude::*;

macro_rules! class_hierarchy {
    ($parent:ident: $($child:ident),*) => {
        paste! {
            #[pyclass(subclass, name=stringify!($parent))]
            pub struct [<Py $parent>];

            $(
                #[pyclass(extends=[<Py $parent>], name=stringify!($child))]
                pub struct [<Py $child>];
            )*
        }
    };
}

class_hierarchy!(Parent: Child1, Child2);
```

Where in this case I would end up with structs named `PyParent`, `PyChild1`, & `PyChild2`, but which correspond to python classes just named `Parent`, `Child1`, and `Child2`, respectively. But this errors because of the `stringify!` usage with "expected string literal".

Is there some way to accomplish what I'm trying to do here? Or is there no way to get around the fact that the `#[pyclass(name)]` *must* be a literal

**Comment by davidhewitt** [comment]
I think it needed to be a literal in the past due to limitations about creating c-strings at compile time. Potentially it would be ok to relax this to any expression which can evaluate at const-time to a c-str. I would be happy to review a PR which experiments with changing the name attribute to accept either a string literal OR an expression which evaluates to `&'static CStr`. I think we'd soon learn if that's actually feasible.
